cc/gtp_player.cc

```diff
@@ -0,0 +1,324 @@
+  auto leaves = MctsPlayer::TreeSearch(batch_size);
+  if (!leaves.empty() && report_search_interval_ != absl::ZeroDuration()) {
+    auto now = absl::Now();
+    if (now - last_report_time_ > report_search_interval_) {
+      last_report_time_ = now;
+      ReportSearchStatus(leaves.back());
+    }
+  }
+  return leaves;
+}
+
+GtpPlayer::Response GtpPlayer::CheckArgsExact(
+    absl::string_view cmd, size_t expected_num_args,
+    const std::vector<absl::string_view>& args) {
+  if (args.size() != expected_num_args) {
+    return Response::Error("expected ", expected_num_args,
+                           " args for GTP command ", cmd, ", got ", args.size(),
+                           " args: ", absl::StrJoin(args, " "));
+  }
+  return Response::Ok();
+}
+
+GtpPlayer::Response GtpPlayer::CheckArgsRange(
+    absl::string_view cmd, size_t expected_min_args, size_t expected_max_args,
+    const std::vector<absl::string_view>& args) {
+  if (args.size() < expected_min_args || args.size() > expected_max_args) {
+    return Response::Error("expected between ", expected_min_args, " and ",
+                           expected_max_args, " args for GTP command ", cmd,
+                           ", got ", args.size(),
+                           " args: ", absl::StrJoin(args, " "));
+  }
+  return Response::Ok();
+}
+
+GtpPlayer::Response GtpPlayer::DispatchCmd(
+    absl::string_view cmd, const std::vector<absl::string_view>& args) {
+  if (cmd == "clear_board") {
+    return HandleClearBoard(cmd, args);
+  } else if (cmd == "echo") {
+    return HandleEcho(cmd, args);
+  } else if (cmd == "final_score") {
+    return HandleFinalScore(cmd, args);
+  } else if (cmd == "gamestate" || cmd == "mg_gamestate") {
+    return HandleGamestate(cmd, args);
+  } else if (cmd == "genmove" || cmd == "mg_genmove") {
+    return HandleGenmove(cmd, args);
+  } else if (cmd == "name") {
+    return HandleName(cmd, args);
+  } else if (cmd == "play") {
+    return HandlePlay(cmd, args);
+  } else if (cmd == "readouts") {
+    return HandleReadouts(cmd, args);
+  } else if (cmd == "report_search_interval") {
+    return HandleReportSearchInterval(cmd, args);
+  }
+  return Response::Error("unknown command");
+}
+
+GtpPlayer::Response GtpPlayer::HandleClearBoard(
+    absl::string_view cmd, const std::vector<absl::string_view>& args) {
+  auto response = CheckArgsExact(cmd, 0, args);
+  if (!response.ok) {
+    return response;
+  }
+
+  NewGame();
+
+  return Response::Ok();
+}
+
+GtpPlayer::Response GtpPlayer::HandleEcho(
+    absl::string_view cmd, const std::vector<absl::string_view>& args) {
+  return Response::Ok(absl::StrJoin(args, " "));
+}
+
+GtpPlayer::Response GtpPlayer::HandleFinalScore(
+    absl::string_view cmd, const std::vector<absl::string_view>& args) {
+  auto response = CheckArgsExact(cmd, 0, args);
+  if (!response.ok) {
+    return response;
+  }
+  if (!game_over()) {
+    return Response::Error("game isn't over");
```

Comment:
For unfinished games, i think correct behavior is to just return the score of the board.  Various GTP harnesses may depend on this behavior.

(Correct failure behavior per spec is to fail with the message "cannot score")

Reply:
done